### PR TITLE
Lua loader: support to load lua files in a encrypted zip file

### PR DESCRIPTION
**Pull request by SunLightJuly**
功能：

使得Lua loader能够载入一个代码包，此代码包里的代码文件经过了字节码编译处理并打成zip包，并有可能经过了xxtea加密处理。

### CODE REVIEW

File: cocos/scripting/lua-bindings/manual/CCLuaStack.cpp [hunk 1]
```diff
@@ -707,6 +707,122 @@ void LuaStack::cleanupXXTEAKeyAndSign()
...
+    if (_xxteaEnabled && _xxteaSign) {
+        if (len) {
+            *len = _xxteaSignLen;
+        }
+        return _xxteaSign;
+    }
+    return nullptr;
+}
+
+static LuaStack *curStack = NULL;
```
Comment (samuele3hu) [inline cocos/scripting/lua-bindings/manual/CCLuaStack.cpp]: Why does it need a static var?

**Comment by samuele3hu** [comment]
Can you push the files related external to https://github.com/cocos2d/cocos2d-x-3rd-party-libs-bin?Thanks.

File: cocos/scripting/lua-bindings/manual/CCLuaStack.cpp (hunk 1 above)
Comment (SunLightJuly) [inline cocos/scripting/lua-bindings/manual/CCLuaStack.cpp]: This static var is only used in the static function "lua_loadChunksFromZIP", and the function was exported to lua.

**Comment by SunLightJuly** [comment]
OK

File: cocos/base/ZipUtils.h [hunk 2]
```diff
@@ -270,7 +271,18 @@ namespace cocos2d
         */
         unsigned char *getFileData(const std::string &fileName, ssize_t *size);
 
+        const std::string getFirstFilename(void);
+        const std::string getNextFilename(void);
```
Comment (minggo) [inline cocos/base/ZipUtils.h]: Please remove `void`.

File: cocos/base/ZipUtils.cpp [hunk 3]
```diff
@@ -612,4 +631,45 @@ unsigned char *ZipFile::getFileData(const std::string &fileName, ssize_t *size)
...
+    std::string path;
+    unz_file_info info;
+    getCurrentFileInfo(&path, &info);
+    return path;
+}
+
+int ZipFile::getCurrentFileInfo(std::string *filename, unz_file_info *info)
+{
+    char path[FILENAME_MAX + 1];
+    int ret = unzGetCurrentFileInfo(_data->zipFile, info, path, sizeof(path), NULL, 0, NULL, 0);
```
Comment (minggo) [inline cocos/base/ZipUtils.cpp]: NULL -> nullptr

**Comment by minggo** [comment]
And please don't add files in external/unzip.
They are designed not to added into source control.

**Comment by minggo** [comment]
@samuele3hu 
Will you send a new pull request based on this PR if you think it is needed.
Thanks.

**Comment by SunLightJuly** [comment]
Very sorry for my busyness and delay.  I will send a new pull request soon. And, still a question here: The added files in external/unzip are parts of unzip module; should I push the files to  https://github.com/cocos2d/cocos2d-x-3rd-party-libs-bin ?

**Comment by SunLightJuly** [comment]
This PR has been renewed. And the files in external/unzip has been pushed to: https://github.com/cocos2d/cocos2d-x-3rd-party-libs-bin/pull/16

**Comment by minggo** [comment]
@SunLightJuly 
Sorry for late reply, i will ask @samuele3hu to review it.
Thanks.

**Pull request closed without merging**